### ISSUE TITLE
Automatic import management

### ISSUE THREAD

**Issue by bsutton** [Dart-Code/Dart-Code#1985]
Coming from an Eclipse/Java environment to a visualcode/dart environment I'm finding managing imports a real pain in the arse.

In Eclipse we have an option 're-organise imports'.

This essentially starts a simple wizard that updates all of your import statements.

It removes unused and invalid imports.
It inserts all required imports.

If it needs assistance to disambiguate an import it asks the user to choose from a set of likely imports (as the the vscode quick fix does).

The wizard also includes a 'skip' button if you are uncertain which import you need. This is important as it lets you resolve the bulk of the imports and lets you go back and manually sort out the odd ones.

Eclipse also allows you to run the same process from the file explorer tree essentially re-organising all of your imports in one go.

In Eclipse we constantly use this feature and it would really make vscode more pleasurable to use. 

In my migration from eclipse to vscode/dart this has to be the biggest pain point.

**Comment by DanTup** [comment]
VS Code does have an "Organize Imports" command which cleans up the imports that are there (removes unused, etc.).

Inserting imports can be done by the code completion - it will automatically add the import when you select an item that isn't already imported.

So, I'm not sure what is really missing. I'm not sure having a way to imsert all missing imports would be that useful (it doesn't seem like there is a common workflow that should end up with a file missing many imports?). I'm also not sure it would work well in VS Code - we can't draw arbitrary dialogs and it's designed to be "an editor", so having wizards feels a little out of scope.

**Comment by bsutton** [comment]
So the core use case is refactoring of files.

Either renaming a library or moving it into a new directory.
I see both of these as critical aspect of good code management.

Both of the above operations can mean that hundreds of files need to be updated to reference the new filename or path.

A secondary use case is when you cut and paste code from another file.
In Eclipse the imports get dropped in automatically.  Whilst that would be my preferred option the request upgraded to organise imports would be sufficient to ease the pain of these very common operations.

I should note that 'organise imports' doesn't remove bad import files.
To be honest Organise imports doesn't do much organising and quick fix is tedious if you have a significant number of imports to fix.

The lack of import management makes any refactoring of file names a real problem.
Good file naming and project tree conventions is critical to good code management.
The current lack of import tools discourages users from refactoring filenames which will result in the poor programing practices.

I'm not quite certain what you mean when you say that vscode is meant to be an editor?
Its the most IDE like editor I've ever used :)

All the developers at Noojee have just moved from java/eclipse to dart/vscode and everyone is bitching about this being a pain point.

Its possibly one of those features that you don't realise how important it is until you suddenly don't have it.

**Comment by DanTup** [comment]
Thanks for the info - makes more sense now. There are some features I'd like that I think solve some of the problems, but they're blocked on VS Code:

- https://github.com/microsoft/vscode/issues/43768 - allows us to update references when moving files
- https://github.com/microsoft/vscode/issues/30066 - allows us to automatically insert the correct imports when you copy/paste code across files

(please add 👍 's to them if you'd like to see them)

> I'm not quite certain what you mean when you say that vscode is meant to be an editor?
> Its the most IDE like editor I've ever used :)

heh, yeah - it's them that have said this, but I agree with you - I've never seen an editor with this sort of functionality! That said, it doesn't have oodles of wizards and dialogs, it doesn't even support a concept of "creating a new project", so I think there're still some way off something like Visual Studio.

Anyway, the real point was that we can't just show custom dialogs. The APIs are (deliberately) very restrictive, so the question isn't always just whether a feature would be good, but also whether we can fit it into their APIs in a way that wouldn't be a terrible user experience.

For example, let's say  we added a command for "insert all missing imports" but there were 15 imports that were ambiguous. There isn't really a good way we can show the user 15 items and let them select from a set for each one. The usual input method is via the command palette, and it's really "one question at a time", so you'd end up this long sequence of selecting imports.

Of course, that might be an edge case - just importing those without conflicts might help in the huge majority of cases. The refactoring in VS Code is all handled by the analyzer from the SDK (the extension doesn't directly parse Dart code, so it can't do this sort of work), thus this would need to be supported there to work well (while we could invoke all of the import quick-fixes sequentially, we'd have to wait for analysis in between every step so it wouldn't be a good experience). There are some open issues I think are related:

- https://github.com/dart-lang/sdk/issues/30296 - supports adding imports in the Organize Imports command
- https://github.com/dart-lang/sdk/issues/37228 - support for "fix all"

Again, it's worth putting 👍 's on these.

**Comment by bsutton** [comment]
I've gone through and thumbed each of the above.

Looks like this area needs a chunk of work.

The lack of being able to display a dialog is a pain.

**Comment by DanTup** [comment]
Tidying up the issue tracker. I think everything mentioned above is covered by other existing issues (and some things - like being able to fix-all-on-save are supported with upcoming SDKs - eg. https://github.com/dart-lang/sdk/commit/85d455fbad6eb6f567e76a9a3726aa19ec8398d4).

If there's anything you think doesn't have another specific issue, let me know, otherwise I'll close this one in favour of those.
